src/referenced.cc

```diff
@@ -157,35 +193,46 @@ bool Referenced::Signature(const Attributes& attributes,
   return true;
 }
 
-std::string Referenced::Hash() const {
-  MD5 hasher;
-  // Use std::set to sort the words
-  std::set<std::string> sorted_absence_keys(absence_keys_.begin(),
-                                            absence_keys_.end());
-  for (const auto& key : sorted_absence_keys) {
-    hasher.Update(key);
+void updateHash(MD5 &hasher, const std::vector<AttributeRef> &keys) {
```

Comment:
* non class member functions should move into annoymous namespace
* UpdateHash 
* hasher is output,  should use raw pointer, move it to be the last argument